File: labs/py/07_tracing.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""
Task 07 — Tracing: Instrument Cora's agent execution with Azure Monitor OpenTelemetry.

Enables telemetry, creates a tracer, runs agent operations with custom spans,
and sends trace data to Application Insights.

Prerequisites:
    - Set TELEMETRY_CONNECTION_STRING in .env (Application Insights connection string).

Run:
    python 07_tracing.py
"""

import os
from dotenv import load_dotenv

load_dotenv()

TELEMETRY_CONNECTION_STRING = os.environ.get("TELEMETRY_CONNECTION_STRING", "")

if not TELEMETRY_CONNECTION_STRING:
    print(
        "❌ TELEMETRY_CONNECTION_STRING is not set in .env.\n"
        "   This script requires an Application Insights connection string.\n"
        "   Find it in Azure Portal → Application Insights → Overview → Connection String.\n"
        "   Then add it to your .env file."
    )
    raise SystemExit(1)

# Initialize Azure Monitor BEFORE importing SDK clients
from azure.monitor.opentelemetry import configure_azure_monitor

configure_azure_monitor(connection_string=TELEMETRY_CONNECTION_STRING)

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode
from azure.ai.projects import AIProjectClient
from azure.ai.projects.models import PromptAgentDefinition
from azure.identity import DefaultAzureCredential

project_endpoint = os.environ.get("AZURE_AI_PROJECT_ENDPOINT", "<project endpoint>")
deployment_name = os.environ.get("MODEL_DEPLOYMENT_NAME", "<model deployment name>")


def main() -> None:
    tracer = trace.get_tracer("cora-zava-diy", "1.0.0")
    credential = DefaultAzureCredential()
    project = AIProjectClient(endpoint=project_endpoint, credential=credential)

    print("📡 Tracing enabled — sending telemetry to Application Insights…\n")

    # Trace agent creation
    with tracer.start_as_current_span(
        "cora.agent.create",
        kind=SpanKind.CLIENT,
        attributes={
            "gen_ai.system": "az.ai.agents",
            "gen_ai.provider.name": "microsoft.agents",
            "model": deployment_name,
            "projectEndpoint": project_endpoint,
        },
    ) as parent_span:
        try:
            # Create the agent within a traced span
            with tracer.start_as_current_span(
                "agents.createVersion",
                kind=SpanKind.CLIENT,
                attributes={"model": deployment_name, "instructions": "You are Cora, Zava DIY assistant"},
            ) as agent_span:
                agent = project.agents.create_version(
                    agent_name="cora-traced-agent",
                    definition=PromptAgentDefinition(
                        model=deployment_name,
                        instructions="You are Cora, the friendly AI assistant for Zava DIY home improvement store.",
                    ),
                )
                agent_span.set_attribute("agent.version", agent.version)
                agent_span.set_attribute("agent.id", agent.id)
                agent_span.set_attribute("agent.name", agent.name)
                agent_span.set_status(StatusCode.OK)
                print(f"✅ Agent created (name: {agent.name}, version: {agent.version})")

            # Trace a test inference
            with tracer.start_as_current_span(
                "cora.inference",
                kind=SpanKind.CLIENT,
                attributes={"query": "What paint for my bathroom?"},
            ) as inference_span:
                openai_client = project.get_openai_client()
                response = openai_client.responses.create(
                    model=deployment_name,
                    instructions="You are Cora, a helpful Zava DIY assistant.",
                    input="What paint should I use for my bathroom?",
                )
                inference_span.set_attribute("response.length", len(response.output_text))
                inference_span.set_status(StatusCode.OK)
                print(f"\n🤖 Response: {response.output_text[:200]}…")

            # The traced agent is preserved for review — it will be cleaned up in Task 9 (teardown).
            print("\n💡 Traced agent preserved. View it in the Foundry Portal under Agents.")

            parent_span.set_status(StatusCode.OK)
        except Exception as e:
            parent_span.set_status(StatusCode.ERROR, str(e))
            parent_span.record_exception(e)
            raise

    print("\n📡 Telemetry sent successfully!")
    print("   View traces in Azure Portal → Application Insights → Transaction search")
    print('   Filter by: customDimensions["gen_ai.system"] = "az.ai.agents"')


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print(f"The sample encountered an error: {err}")

File: labs/js/src/07-tracing.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * Task 07 — Tracing: Instrument Cora's agent execution with Azure Monitor OpenTelemetry.
 *
 * This script enables telemetry, creates a tracer, runs agent operations
 * with custom spans, and sends trace data to Application Insights.
 *
 * @summary Trace agent execution with Azure Monitor OpenTelemetry.
 *
 * Prerequisites:
 *   - Set TELEMETRY_CONNECTION_STRING in .env (Application Insights connection string).
 *
 * Run:
 *   npm run build && node dist/07-tracing.js
 */

import {
  useAzureMonitor,
  shutdownAzureMonitor,
} from "@azure/monitor-opentelemetry";
import type { AzureMonitorOpenTelemetryOptions } from "@azure/monitor-opentelemetry";
import { SpanKind } from "@opentelemetry/api";
import "dotenv/config";

const TELEMETRY_CONNECTION_STRING =
  process.env["TELEMETRY_CONNECTION_STRING"] || "";

if (!TELEMETRY_CONNECTION_STRING) {
  console.error(
    "❌ TELEMETRY_CONNECTION_STRING is not set in .env.\n" +
    "   This script requires an Application Insights connection string.\n" +
    "   Find it in Azure Portal → Application Insights → Overview → Connection String.\n" +
    "   Then add it to your .env file.",
  );
  process.exit(1);
}

const options: AzureMonitorOpenTelemetryOptions = {
  azureMonitorExporterOptions: {
    connectionString: TELEMETRY_CONNECTION_STRING,
  },
};

// Initialize Azure Monitor BEFORE importing SDK clients
useAzureMonitor(options);

import { trace, SpanStatusCode } from "@opentelemetry/api";
import { AIProjectClient } from "@azure/ai-projects";
import { DefaultAzureCredential } from "@azure/identity";

const projectEndpoint =
  process.env["AZURE_AI_PROJECT_ENDPOINT"] || "<project endpoint>";
const deploymentName =
  process.env["MODEL_DEPLOYMENT_NAME"] || "<model deployment name>";

export async function main(): Promise<void> {
  const tracer = trace.getTracer("cora-zava-diy", "1.0.0");
  const project = new AIProjectClient(
    projectEndpoint,
    new DefaultAzureCredential(),
  );

  console.log("📡 Tracing enabled — sending telemetry to Application Insights…\n");

  // Trace agent creation
  await tracer.startActiveSpan(
    "cora.agent.create",
    {
      kind: SpanKind.CLIENT,
      attributes: {
        "gen_ai.system": "az.ai.agents",
        "gen_ai.provider.name": "microsoft.agents",
        model: deploymentName,
        "http.method": "POST",
        projectEndpoint: projectEndpoint,
      },
    },
    async (span) => {
      try {
        // Create the agent within a traced span
        await tracer.startActiveSpan(
          "agents.createVersion",
          { kind: SpanKind.CLIENT },
          async (agentSpan) => {
            try {
              agentSpan.setAttribute("deploymentName", deploymentName);
              agentSpan.setAttribute("model", deploymentName);
              agentSpan.setAttribute(
                "instructions",
                "You are Cora, Zava DIY assistant",
              );

              const res = await project.agents.createVersion(
                "cora-traced-agent",
                {
                  kind: "prompt",
                  model: deploymentName,
                  instructions:
                    "You are Cora, the friendly AI assistant for Zava DIY home improvement store.",
                },
              );

              agentSpan.setAttribute("agent.version", res.version);
              agentSpan.setAttribute("agent.id", res.id);
              agentSpan.setAttribute("agent.name", res.name);
              agentSpan.setStatus({ code: SpanStatusCode.OK });

              console.log(`✅ Agent created (name: ${res.name}, version: ${res.version})`);
              return res;
            } catch (e) {
              agentSpan.setStatus({
                code: SpanStatusCode.ERROR,
                message: String(e),
              });
              throw e;
            } finally {
              agentSpan.end();
            }
          },
        );

        // Trace a test inference
        await tracer.startActiveSpan(
          "cora.inference",
          { kind: SpanKind.CLIENT },
          async (inferenceSpan) => {
            try {
              const openAIClient = await project.getOpenAIClient();
              inferenceSpan.setAttribute("query", "What paint for my bathroom?");

              const response = await openAIClient.responses.create({
                model: deploymentName,
                instructions: "You are Cora, a helpful Zava DIY assistant.",
                input: "What paint should I use for my bathroom?",
              });

              inferenceSpan.setAttribute(
                "response.length",
                response.output_text.length,
              );
              inferenceSpan.setStatus({ code: SpanStatusCode.OK });

              console.log(`\n🤖 Response: ${response.output_text.substring(0, 200)}…`);
            } catch (e) {
              inferenceSpan.setStatus({
                code: SpanStatusCode.ERROR,
                message: String(e),
              });
              throw e;
            } finally {
              inferenceSpan.end();
            }
          },
        );

        // The traced agent is preserved for review — it will be cleaned up in Task 9 (teardown).
        console.log("\n💡 Traced agent preserved. View it in the Foundry Portal under Agents.");

        span.setAttribute("http.status_code", 200);
        span.setStatus({ code: SpanStatusCode.OK });
      } catch (e) {
        span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
        span.recordException(e as Error);
        span.setAttribute("http.status_code", 500);
        throw e;
      } finally {
        span.end();
      }
    },
  );

  console.log("\n📡 Telemetry sent successfully!");
  console.log(
    "   View traces in Azure Portal → Application Insights → Transaction search",
  );
  console.log(
    '   Filter by: customDimensions["gen_ai.system"] = "az.ai.agents"',
  );

  await shutdownAzureMonitor();
}

main().catch(async (err) => {
  console.error("The sample encountered an error:", err);

  // Flush any pending telemetry before shutdown
  const provider = (trace as any).getTracerProvider?.();
  if (provider && typeof provider.forceFlush === "function") {
    await provider.forceFlush();
  }
  await new Promise((resolve) => setTimeout(resolve, 5000));

  await shutdownAzureMonitor();
});
